File: java/src/main/java/org/libpetri/core/internal/OutputActionCheck.java
package org.libpetri.core.internal;

import org.libpetri.core.PetriNet;
import org.libpetri.core.Transition;
import org.libpetri.core.TransitionAction;

/**
 * CORE-043 reconciliation of declared structure against bound behaviour, shared by the
 * execution and verification entry points.
 */
public final class OutputActionCheck {

    private OutputActionCheck() {}

    /**
     * Rejects any transition that declares an output spec while carrying the built-in
     * {@link TransitionAction#passthrough()} — passthrough produces no tokens, so every firing
     * would fail output validation (IO-015) and the declared output would never arrive.
     *
     * @param net the net about to be compiled or verified
     * @throws IllegalStateException on the first such transition, naming it
     */
    public static void requireOutputProducingActions(PetriNet net) {
        for (var t : net.transitions()) {
            requireOutputProducingAction(t);
        }
    }

    /** Single-transition form of {@link #requireOutputProducingActions(PetriNet)}. */
    public static void requireOutputProducingAction(Transition t) {
        if (t.outputSpec() != null && TransitionAction.isPassthrough(t.action())) {
            throw new IllegalStateException(
                ("Transition '%s' declares an output spec but carries passthrough(), which "
                    + "produces no tokens. Every firing would fail output validation (IO-015) and "
                    + "the declared output would never arrive. Bind an action that produces it — "
                    + "TransitionAction.fork() moves the input token across — or drop the output "
                    + "spec if the transition is meant to be a sink.").formatted(t.name()));
        }
    }
}

File: typescript/src/core/internal/output-action-check.ts
import type { PetriNet } from '../petri-net.js';
import { isPassthrough } from '../transition-action.js';

/**
 * CORE-043: a transition that declares an output spec must not carry the built-in
 * `passthrough()`. It produces no tokens, so output validation (IO-015) rejects every
 * firing and the declared output never arrives. Enforced when a net is compiled for
 * execution and when one is handed to verification, so verification cannot green-light a net that will not compile.
 */
export function requireOutputProducingActions(net: PetriNet): void {
  for (const t of net.transitions) {
    if (t.outputSpec !== null && isPassthrough(t.action)) {
      throw new Error(
        `Transition '${t.name}' declares an output spec but carries passthrough(), which ` +
        `produces no tokens. Every firing would fail output validation (IO-015) and ` +
        `the declared output would never arrive. Bind an action that produces it — ` +
        `fork() moves the input token across — or drop the output spec if the ` +
        `transition is meant to be a sink.`,
      );
    }
  }
}
